### ISSUE TITLE
Security, is it a principle or best practice?

### ISSUE THREAD

**Issue by Beanow** [sfosc/sfosc#85]
Over in #33 the suggestion was to add a principle:
> * All software must be developed with security in mind to avoid introducing vulnerabilities across the open source ecosystem.

There I did comment https://github.com/sfosc/sfosc/pull/33#issuecomment-490926053 for Rawls idea of justice I don't think you don't need to develop for security, but should have a responsible disclosure.

However I think more important to figure out is what to do with the sentiment that security is important for sustainable free and open source communities.

It makes intuitive sense that better security would be helpful. Then again so would better tutorials or being more friendly. Does security have a special dynamic we need to address?

@GeorgLink made a really interesting observation https://github.com/sfosc/sfosc/pull/33#issuecomment-448977625 on that I think. We do have a code of conduct in the principles for protection of the community:
> * We must have a strong code of conduct, with clear, fair enforcement mechanisms.

**Comment by nothingismagick** [comment]
As a member of the security community, I just want to make a note in this
discussion that security is everyone's responsibility. However, in the
sense of sustainability of open source projects and especially libraries
for downstream consumption, it should be a requirement that dependencies
are kept up to date and the underlying code is secure.

**Comment by nothingismagick** [comment]
I'm not sure exactly how best to formulate this concern, but I'll think
about it over the weekend and make a proposal.

**Comment by Beanow** [comment]
@nothingismagick that certainly makes sense. But isn't this a case of making the software fit-for-purpose?

For example if you have defined an API with lots of downstream clients using it. You can't break it or bad things will happen. I don't feel like we need to add this (API) scenario to SFOSC principles or rules.

**Comment by nothingismagick** [comment]
Well, downstream is just the penumbra of emanations. Being dedicated to
security for its own sake IS a factor in the sustainability of a project. I
mean, today, when I see a project that has out of date dependencies or is
flagged as insecure I will either make a PR to inform the project owner or
just look for a project that is regularly maintained.

For me, receptivity to security concerns really says a lot about the
community, and passing the buck or ignoring the problem is a huge warning.

**Comment by Beanow** [comment]
To play devils advocate a little more here:

A generally receptive community for any valid concerns is a critical factor for sustainability. Security issues or broken APIs are both valid concerns.

Openly being committed to security won't hurt. But neither would committing to API compatibility. 

An ignored security issue looks bad. But so does an ignored broken API issue.

Based on my earlier train of thought. Someone with commit & deploy access of a project has a position that others do not. Them being dedicated to a responsible disclosure process to fairly distribute the ability to fix sensitive issues responsibly seems like a step forward.
This may be different for API compatibility, because downgrading (or not upgrading) to earlier versions would solve your problem. So in that case commit and release histories would be the saving grace. While a security bug may put any range of earlier versions at risk.

So maybe this "it can break things in retrospect" is what makes security unique?

**Comment by nothingismagick** [comment]
Yes. There might be modules out there that are perfectly written, have zero
dependencies and will never have security issues. So in this highly
unlikely case, we could theoretically ignore the sustainability of security.

For the rest of us who are human, sustainably maintaining software means
being aware of vulnerabilities in underlying libraries, operating systems
and human behaviours.

**Comment by Beanow** [comment]
Right. So speaking of dependencies and dependents. Especially in the case of libraries. Security problems propagate to dependents and may mean you have a shared responsibility towards the ecosystem at large. Even when you didn't change the code at all.

This networking effect may negatively impact sustainability of other open source projects, so an SFOSC should follow security best practices?

**Comment by nothingismagick** [comment]
Exactly.

**Comment by Beanow** [comment]
Well then, the original thought of putting effort into security while designing the software is important. As an md5 passwords project gaining more users would be detremental.

But on top of that response in a responsible disclosure way is important. *And* I would argue publication is a critical part of the response. I lost count of the times `npm audit` told me about vulnerabilities I never would have found otherwise.

cc @GeorgLink, what is your take? Do these three aspects cover your security concerns?

**Comment by themightychris** [comment]
I think we address much of the practical surface area for making security sustainable with the basic sustainability tests:

- Can a non-core developer upgrade all dependencies?
- Can a non-core developer make major internal changes and release them with confidence?

I suspect that covers most of the security issues we can get at with codified practices, preventing situations where downstream users/maintainers are afraid to touch dated dependencies. 

I also like the idea of requiring projects to contain policies on reporting security issues and disclosures, and having model templates folks can use

**Comment by GeorgLink** [comment]
@Beanow:

> cc @GeorgLink, what is your take? Do these three aspects cover your security concerns?

I appreciate that the conversation is occurring - that was the goal of my initial proposal. I am not an expert on security like others here in the discussion and will defer to their opinion.

**Comment by Beanow** [comment]
@GeorgLink fair enough. Thanks again for getting this started.

@themightychris those are good points. Making it easier to help out or do it yourself after a fork. But I also think the commitment of core developers is important. Even if access to the privileged position is broad and fair, through an election maybe, in my opinion the current holders of the position should commit to the three aspects I got to in an earlier comment.

**Comment by nothingismagick** [comment]
@lirantal - do you think you might be able to help us uncover the best most appropriate way to describe responsible disclosure?

**Comment by lirantal** [comment]
Thanks for looping me in @nothingismagick, I'm happy to join the conversation. a small disclaimer that I didn't read yet about SFOSC so I don't have yet the full context but will catch up on this soon enough.

I'm excited to see the discussion here about a responsible security disclosure and the whole notion of security awareness and commitment for open source projects.

I created a generic template for responsible disclosure that projects can adopt (https://github.com/nodejs/security-wg/blob/master/processes/responsible_disclosure_template.md) and can use as a reference point and that we also hope to increase awareness for it through a badge (see the badges on https://github.com/nodejs/security-wg). 

Moreover, through my community engagement work at Snyk I have recently worked with the verdaccio team to establish and promote security guidelines and practices for that project for which we have created the following more elaborate and project-specific security disclosure: https://github.com/verdaccio/verdaccio/blob/master/SECURITY.md 

Besides taking verdaccio's example and replicating it, I'm happy to see how I can expand this community engagement I have going right now to more areas in SFOSC if that's relevant.

**Comment by Beanow** [comment]
Those are awesome @lirantal 
Initially looking at the security-wg template, I was a little overwhelmed by the reference to [processes/third_party_vuln_process.md](https://github.com/nodejs/security-wg/blob/master/processes/third_party_vuln_process.md) as that looks like some seriously demanding terms for a small project. With the verdaccio example though things made a lot more sense.

I've got two burning questions,

One is, do you think these three aspects cover everything related to protecting the larger oss ecosystem or am I missing something?

- Design and develop the software for security.
  _Example, don't let MD5 password hashing get merged because it "isn't that important for this project"._
- Prioritize and respond appropriately to vulnerability reports.
  _The responsible disclosure process goes here._
- Openly publish vulnerability reports.
  _Example, npm audit style. Whatever is most effective at informing your users / dependents._

The second is, I would like to point to (or if need be, create) a resource explaining the importance and providing a guide on how to do it properly. Something along the lines of https://keepachangelog.com/en/1.0.0/
Can you recommend a good resource, or help create something along these lines?

**Comment by lirantal** [comment]
Happy to hear it helps! :-)

About the first question - those are great points to start or work on but that's probably not nearly an encompassing and comprehensive list. I can think of so many other things for example compromised user accounts and how do you raise awareness and provide tooling to help developers keep their token secrets for example? What about threat modeling the actual supply chain? DevSecOps efforts such as security testing in a CI build to prevent from security issues going through?

About the changelog example - I'm not entirely sure what you had in mind but it may be connected to an agenda that I had started working on already with regards to open source security maturity models which I mentioned in my last comment. I'm happy to connect you with on that if it's an initiative you'd like to get involved with.

**Comment by nothingismagick** [comment]
We'd love to discuss this with you - perhaps join our discord server and we
can set up a time next week that works for all of us.

https://discord.gg/PxFSTU6

**Comment by lirantal** [comment]
Thanks for the invite 🙏

I will do my best to join but will probably find it hard to catch up in synchronous chats. I'll pop-up on discord in a couple of weeks and see how it works. Regardless, you can always ping me, or mention me and I'll be happy to chime in and help where I can.
